File: 278_first-bad-version_bi-search.cpp
// The API isBadVersion is defined for you.
// bool isBadVersion(int version);

class Solution {
public:
    int firstBadVersion(int n) {
        int low = 1, high = n, mid;
        while(low <= high){
            mid = low + (high - low) / 2;
            if(isBadVersion(mid))
                high = mid - 1;
            else
                low = mid + 1;
        }
        // The loop terminate when low is larger than high
        // and low will point at the first bad version here
        return low;
    }
};

File: 278_first-bad-version_bi-search.py
# The isBadVersion API is already defined for you.
# def isBadVersion(version: int) -> bool:

class Solution:
    def firstBadVersion(self, n: int) -> int:
        low = 1; high = n
        while low < high:
            mid = (low + high) // 2
            if isBadVersion(mid):
                # Our candidate will still include mid
                high = mid
            else:
                # Our candidate won't include mid
                low = mid + 1
        return low
